File: src/CardGame/Game.hs
module CardGame.Game (
    GameState (..)
    , Game (..)
    , dealCards
    , createEmptyGame
    , gameActions
) where


import CardGame.Player ( deal, Player )
import CardGame.Card ( Card )

import Data.CircularList ( CList, empty )
import Feature (Feature)
import CardGame.PlayerMove (Move)
import CDSL.CDSLExpr (CDSLExpr)

data GameState = Start | TurnStart | TurnEnd | RoundStart | RoundEnd | End
    deriving (Show, Eq)

data Game = Game {
    gameName :: String
    , playerMoves :: [(Move, Bool)]
    , cardGen :: [Card]
    , discard :: [Card]
    , turnOrder :: [CDSLExpr]
    , cardSuits :: [CDSLExpr]
    , cardRanks :: [CDSLExpr]
    , cardEffects :: [CDSLExpr]
    , deck :: [Card]
    , pile :: [(Card, Maybe Card)]
    , players :: CList Player
    , endCon :: [Game -> Bool]
    , winCon :: [Game -> [Player]]
    , state :: GameState
    , rules :: [(Feature, [CDSLExpr])]
    , actions :: [(GameState, [(Game -> IO Game, Bool)])]
    , rounds :: Int
    , canPlaceCard :: [Game -> Card -> Bool]
    , cpcException :: [CDSLExpr]
}



dealCards :: Game -> Int -> Game
dealCards game n = do
    let (plrs', deck') = deal n (deck game) (players game)
    game { players = plrs', deck = deck'}


createEmptyGame :: Game
createEmptyGame = Game {
                    gameName = "Default Game"
                    , turnOrder = []
                    , discard = []
                    , cpcException = []
                    , playerMoves = []
                    , cardGen = []
                    , cardSuits = []
                    , cardRanks = [] 
                    , cardEffects = []
                    , deck = []
                    , pile = []
                    , players = empty
                    , endCon = []
                    , winCon = []
                    , state = Start
                    , rules = []
                    , actions = []
                    , rounds = 0
                    , canPlaceCard = []
                    }

gameActions :: [[Game -> IO Game]] -> Game -> IO Game
gameActions xs = go (concat xs)
    where
        go :: [Game -> IO Game] -> Game -> IO Game
        go [] gm = return gm
        go (f:fs) gm = f gm >>= go fs

File: src/CardGame/Game.rs
use super::Player::Player;
use super::Player::Move;
use super::Card::Card;
use crate::Feature::Feature;
use crate::CDSL::Expr::Expr;

#[derive(Clone, Copy, Debug)]
pub enum GameState {
    Start,
    TurnStart,
    TurnEnd,
    RoundStart,
    RoundEnd,
    End
}

pub struct Game {
    game_name: String,
    state: GameState,
    player_moves: Vec<(Move, bool)>,
    card_gen: Vec<Card>,
    discard: Vec<Card>,
    deck: Vec<Card>,
    pile: Vec<(Card, Option<Card>)>,
    players: Vec<Player>,
    rules: Vec<(Feature, Vec<Expr>)>,
    actions: Vec<(GameState, Vec<Expr>)>,
    can_place_card: Vec<String>
}


impl Game {
    pub fn new(name: String, plrs: Vec<String>) -> Game {
        Game {
            game_name: name,
            state: GameState::Start,
            players: Vec::new(),
            player_moves: Vec::new(),
            card_gen: Vec::new(),
            discard: Vec::new(),
            deck: Vec::new(),
            pile: Vec::new(),
            rules: Vec::new(),
            actions: Vec::new(),
            can_place_card: Vec::new()
        }
    }
}
